### PR TITLE
Add the DASDAE inventory model layer

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

This PR adds the core model layer of the DASDAE inventory — a StationXML-style, fiber-native metadata model for DFOS observing systems — as `dascore.core.inventory`. It is the first of several planned PRs (patch enrichment, spool integration, and the directory authoring loader will follow separately).

What it provides:

- Immutable pydantic models for the whole tree: `Inventory`, `Network`, `Station`/`Channel`/`Response`, `FiberArray`, `Acquisition`, `DistanceMap`, `OpticalPath` with its component/geometry/coupling/annotation tracks, and shareable resources (`Interrogator`, `Cable`, `Enclosure`, `ExternalResource`).
- Track rules: optical components tile the path span; geometry and coupling are function tracks (partial coverage is legal and resolves as NaN, overlap raises); annotations overlap freely.
- Half-open UTC validity epochs (`[start, end)`, NaT = ongoing): abutting epochs are legal, overlapping same-code epochs raise at `check()`, and no more than one optical path per `(fiber array, location code)` is valid at a time.
- Channel-to-distance resolution: affine (`start_distance` + `spatial_interval`) or a measured `DistanceMap` (channel or instrument-distance input axis), mutually exclusive.
- `data_source_id` + time resolution (`Inventory.resolve`), code charset enforcement, coordinates stored on canonical `(x, y, z)` axes with CRS-resolved label aliases.
- Shareable resources normalize into the flat `resources` pool with `resource_id` references: inline objects are hoisted at construction, references are type-checked, and `replace()` corrections on a resource are single-site.
- Path operations preserving absolute optical distances (`select`, `split_at`, `reverse`, concatenation), YAML round-trip, and `dc.inventory()`.

Design background: the model was developed through an iterative spec process (with adversarial design reviews) in the DASDAE/inventory repo; that content will migrate into DASCore's docs in a follow-up. The implementation went through two internal review rounds plus two external review rounds; all findings are covered by regression tests in `tests/test_core/test_inventory.py` (82 tests).

Notable shared-machinery changes: `dascore/utils/models.py` gains `InventoryModel`/`TimeRangedModel` bases, and `sensible_model_equals` now compares nested structures recursively with positional null-matching (previously nested NaT/NaN values made structurally identical models compare unequal).

## Changelog

- added: `dascore.core.inventory`, a StationXML-style fiber-native metadata model — `Inventory`, `Network`, `FiberArray`, `Acquisition`, `OpticalPath` and its tracks, plus shareable `Interrogator`/`Cable`/`Enclosure` resources.

## Checklist

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [ ] added the "ready_for_review" tag once the PR is ready to be reviewed.

## Summary by CodeRabbit

* **New Features**
  * Added a comprehensive inventory model for networks, stations, fiber arrays, optical paths, acquisitions, resources, geometry, and coordinate systems.
  * Added inventory validation, resource lookup, time-based resolution, path operations, channel-to-distance mapping, and YAML serialization.
  * Added coordinate interpolation, resource replacement, time-aware metadata handling, and convenient factory creation.
  * Added clear errors for invalid inventory metadata.

* **Tests**
  * Added extensive coverage for inventory construction, validation, mapping, resolution, serialization, equality, and edge cases.

### CODE REVIEW

File: dascore/core/inventory.py [hunk 1]
```diff
@@ -0,0 +1,1600 @@
...
+    version: str = Field(default="", description="Content version of the document.")
+
+
+class CoordinateReferenceSystem(InventoryModel):
+    """
+    Inventory-wide coordinate reference system.
+
+    All coordinate-bearing metadata is interpreted using this CRS. The default
+    is geographic WGS84 3D (EPSG:4979); override it only for exceptional
+    frames such as mines or laboratories.
```
Comment (d-chambers) [inline dascore/core/inventory.py]: We dont need the guidelines; we are all adults here ;)

File: dascore/core/inventory.py [hunk 2]
```diff
@@ -0,0 +1,1600 @@
...
+
+The inventory extends the StationXML concept with first-class support for
+fiber-optic arrays. It describes the physical optical path (fiber, connectors,
+splices), the geometry, coupling, and annotation tracks along optical
+distance, and the interrogator configurations (acquisitions) that produced
+DAS patches. Patches carry a ``data_source_id``
+(``network.fiber_array.location.acquisition``) which, together with time,
+resolves against an inventory.
+
+Key model rules (see the DASDAE inventory specification):
```
Comment (d-chambers) [inline dascore/core/inventory.py]: we can remove the reference to the spec; remember the goal is an implementation, not a spec.

File: dascore/core/inventory.py [hunk 3]
```diff
@@ -0,0 +1,1600 @@
...
+- Optical components are the tiling track: they cover the whole path exactly
+  once. Geometry and coupling are function tracks: coverage may be partial
+  (uncovered distance is undefined), overlap raises. Annotations overlap
+  freely.
+- Codes use letters, digits, and ``-``; ``.`` is the ``data_source_id``
+  separator. All codes are non-empty except ``location_code``.
+- ``Acquisition`` maps channels onto path distance through either the affine
+  form (``start_distance`` + ``spatial_interval``) or a measured
+  ``DistanceMap``; the two are mutually exclusive.
+"""
```
Comment (d-chambers) [inline dascore/core/inventory.py:13]: Let's make this much more concise. Mostly,  such info should live in the appropriate class. You can leave just a simple overview here, not detailed rules. Such rules will be included in a dedicated note in a future PR.

File: dascore/core/inventory.py [hunk 4]
```diff
@@ -0,0 +1,1600 @@
...
+        description="Stable identifier for this shareable inventory resource.",
+    ),
+]
+NotesStr = Annotated[str, Field(default="", description="Additional notes.")]
+
+
+def _is_strictly_increasing(values) -> bool:
+    """Return True if a sequence is strictly increasing."""
+    arr = np.asarray(values, dtype=float)
+    return bool(len(arr) < 2 or np.all(np.diff(arr) > 0))
```
Comment (d-chambers) [inline dascore/core/inventory.py]: This smells like a general util that shouldnt live here (there might already be one in utils).

File: dascore/core/inventory.py [hunk 5]
```diff
@@ -0,0 +1,1600 @@
...
+    code: str = Field(default="4979", description="Authority code for this CRS.")
+    name: str = Field(default="WGS 84 3D", description="Human-readable CRS name.")
+    coordinate_labels: tuple[str, ...] = Field(
+        default=("longitude", "latitude", "elevation"),
+        description=(
+            "Meaning of the canonical (x, y, z) axes, in order. Labels come "
+            "from the controlled coordinate vocabulary."
+        ),
+    )
+    units: str = Field(default="degree", description="Coordinate units when known.")
```
Comment (d-chambers) [inline dascore/core/inventory.py]: This wouldnt apply to elevation. Maybe this should be a tuple (x_units, y_unit, z_unit) instead?

File: dascore/core/inventory.py [hunk 6]
```diff
@@ -0,0 +1,1600 @@
...
+    vertical_datum: str = Field(
+        default="", description="Vertical datum or reference surface, if known."
+    )
+
+    @field_validator("coordinate_labels")
+    @classmethod
+    def _check_labels(cls, value):
```
Comment (d-chambers) [inline dascore/core/inventory.py:193]: Can this be baked into the type annotations rather than having a validator?

File: dascore/core/inventory.py [hunk 7]
```diff
@@ -0,0 +1,1600 @@
...
+    serial_number: str = Field(default="", description="Serial number.")
+    instrument_type: str = Field(
+        default="interrogator", description="General instrument category."
+    )
+
+
+class Enclosure(InventoryModel):
+    """Physical housing, pipe, duct, conduit, or carrier resource."""
+
+    type: Literal["Enclosure"] = "Enclosure"
```
Comment (d-chambers) [inline dascore/core/inventory.py]: Do we need other words? You list some in the docstring but dont support specifying that type here?

File: dascore/core/inventory.py [hunk 8]
```diff
@@ -0,0 +1,1600 @@
...
+
+class Geometry(InventoryModel):
+    """
+    Geometry for an interval of an optical path.
+
+    A geometry is a piecewise segment placed by its ``distance`` array: at
+    least two strictly increasing optical distances, each paired with the
+    coordinate at that point (interpreted using the inventory CRS). Coverage
+    is the half-open span of the array; there is no separate length field. A
+    coil is a segment whose coordinates repeat while distance advances.
```
Comment (d-chambers) [inline dascore/core/inventory.py]: "A coil, or other 'clump'" can be represented... "

File: dascore/core/inventory.py [hunk 9]
```diff
@@ -0,0 +1,1789 @@
...
+    direction: str = Field(
+        default="", description="Measurement direction: forward or reverse."
+    )
+    data: ExternalResource | str | None = Field(
+        default=None, description="The trace or report file, as a resource."
+    )
+
+
+class Interrogator(InventoryModel):
+    """DAS interrogator unit used for data collection."""
```
Comment (d-chambers) [inline dascore/core/inventory.py]: Should say DFOS. We want DASCore to work for more than just DAS

File: dascore/core/inventory.py [hunk 10]
```diff
@@ -0,0 +1,1789 @@
...
+        description=(
+            "Functional enclosure type, such as splice_box, junction_box, "
+            "coupler, turnaround, protective_housing, conduit, pipe, duct, "
+            "casing, borehole, or tray."
+        ),
+    )
+    material: str = Field(default="", description="Material of the enclosure.")
+    manufacturer: str = Field(default="", description="Manufacturer name.")
+    model: str = Field(default="", description="Model name.")
+    inner_diameter: float | None = Field(
```
Comment (d-chambers) [inline dascore/core/inventory.py]: Maybe we just call it diameter, since now we don't have an outer diameter?

File: dascore/utils/models.py [hunk 11]
```diff
@@ -87,16 +95,31 @@ def sensible_model_equals(self: BaseModel | Mapping, other: object) -> bool:
...
-                return False
-        else:
-            if val1 != val2 and not (_all_null(val1) and _all_null(val2)):
-                return False
+        if not _values_equal(d1[name], d2[name]):
+            return False
     return True
 
 
+def _values_equal(val1, val2) -> bool:
```
Comment (d-chambers) [inline dascore/utils/models.py:103]: Doesnt the DASCoreBaseModel already have __eq__ that can be used here?
